### ISSUE TITLE
Policy: REINFORCE (Monte Carlo Policy Gradient)

### ISSUE THREAD

**Issue by espenhk** [espenhk/gamer-ai#21]
## Summary

Add a **REINFORCE policy** (`REINFORCEPolicy`) — a pure Monte Carlo policy gradient over a softmax discrete action distribution and a continuous Gaussian steering head. This is the simplest gradient-based policy and bridges the gap between the current evolutionary/tabular methods and full actor-critic algorithms.

## Motivation

The codebase currently has:
- **Gradient-free** methods: GeneticPolicy, CMA-ES (planned), hill-climbing
- **Tabular TD**: EpsilonGreedy, MCTSPolicy  
- **External gradient**: PPO via SB3 (rl/train.py)

There is no **in-process gradient-based policy** in `policies.py`. REINFORCE fills this gap and serves as a foundation for actor-critic extensions (A2C, PPO-from-scratch).

Key properties:
- No replay buffer needed (on-policy)
- No target network needed
- Pure numpy — no external ML library
- Natural action probability outputs (useful for entropy regularisation later)

## Proposed Implementation

```python
class REINFORCEPolicy(BasePolicy):
    """
    REINFORCE with baseline for variance reduction.
    Action head: softmax over 9 discrete actions (steering abstraction).
    Accumulates (log_prob, reward) pairs within an episode.
    Gradient update on episode end using discounted returns.
    """
```

### Architecture

```
Input: obs (21 + n_lidar_rays dims)
Hidden: [64, 64] (ReLU)
Output: logits for 9 discrete actions  (softmax → action probs)

Baseline: scalar value head sharing first hidden layer
  → reduces gradient variance without bias
```

### Algorithm sketch

```
Each step:
  1. Forward pass: logits = net(obs)
  2. probs = softmax(logits)
  3. Sample action index a ~ Categorical(probs)
  4. Apply action; receive reward r
  5. Store (log_prob(a), r) in episode buffer

Episode end:
  1. Compute discounted returns G_t = Σ_{k≥t} γ^k r_{t+k}
  2. Normalise G_t: (G_t - mean) / (std + ε)  [variance reduction]
  3. Baseline estimate b: running mean of G_t
  4. Policy gradient: ∇θ = Σ_t (G_t - b) * ∇log π(a_t|s_t)
  5. Update: θ ← θ + α * ∇θ
  6. Clear episode buffer
```

### Config additions

```yaml
policy_type: reinforce
policy_params:
  hidden_sizes: [64, 64]
  learning_rate: 0.001
  gamma: 0.99
  entropy_coeff: 0.01     # H(π) bonus to encourage exploration
  baseline: running_mean  # or "none"
```

## Files to change

| File | Change |
|------|--------|
| `policies.py` | Add `REINFORCEPolicy` (~150 lines) |
| `main.py` `_make_policy()` | Register `"reinforce"` case |
| `config/training_params.yaml` | Document policy + params |
| `tests/test_policies.py` | Unit test: gradient sign correct on trivial reward |

## Key Design Decisions

- **Discrete action head over `_DISCRETE_ACTIONS`**: avoids the `accel`/`brake` binary thresholding problem of the continuous head. Gradient flows naturally through the softmax.
- **Running mean baseline**: simple and unbiased; tracks a scalar EMA of past episode returns.
- **Entropy bonus** (`entropy_coeff`): adds H(π) = -Σ p log p to the loss, preventing premature policy collapse, especially useful early in training.
- **Episode-level update** (`on_episode_end`): consistent with existing hook pattern. No partial-episode updates.
- **Serialization**: save weights as nested list (same as `NeuralNetPolicy` YAML format) plus running baseline value.

## Acceptance Criteria

- [ ] `REINFORCEPolicy` inherits `BasePolicy`, implements `__call__`, `update`, `on_episode_end`, `to_cfg`, `save`
- [ ] Gradient update only on episode end; `update()` accumulates log_prob + reward each step
- [ ] Discounted returns normalised before gradient step
- [ ] Entropy regularisation controlled by `entropy_coeff` (0 = off)
- [ ] `policy_type: reinforce` works end-to-end in `main.py`
- [ ] Unit test: gradient ascent increases log-probability of high-reward actions on a 1-step synthetic environment

## Extensions (future)

- A2C: add a separate value network (critic) trained with TD(λ) alongside REINFORCE actor
- PPO-from-scratch: clip the REINFORCE objective ratio π_new/π_old — removes SB3 dependency from `rl/train.py`
